### PR TITLE
committing 3 .ipynb files 2 that belong in the 2.train-VAE module tha…

### PR DESCRIPTION
**Pull request by mjw-18**
…t optimize hyper parameters and generate the training curve graph. the third .ipynb file that was edited was the data split .ipynb. this commit also includes the VAE parameterization .py files that were sourced from cell-health and then repurposed to function with our data. an update to the gitignore file excludes the saved trials generated from the VAE hyperparameterization.

### CODE REVIEW

**Comment by mjw-18** [comment]
@gwaybio

**Comment by gwaybio** [comment]
Got it! I will review this week.

Just checking @mjw-18 Are these the most up to date files? (You had mentioned you needed to check)

**Comment by mjw-18** [comment]
yep these are up to date!

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,5 @@
 0.data-download/data/*
 1.data-exploration/test_data_loader.ipynb
 __pycache__
+2.train-VAE/hyper_parameterization.ipynb
```
Comment (gwaybio) [inline .gitignore]: is this intentional? Should we include this file? Oh, is it actually saved later as a `.py` file (instead of `.ipynb`)?

File: 2.train-VAE/scripts/nbconverted/beta_vae_training.py [hunk 2]
```diff
@@ -29,56 +28,117 @@
...
-dependency_df1.head(3)
+# subsetting the genes 
+# create dataframe containing the 1000 genes with the largest variances and their corresponding gene label and extract the gene labels
+largest_var_df = gene_stats.nlargest(1000, "variance")
+gene_list = largest_var_df["gene_ID"].tolist()
+gene_list
+
+# create new training and testing dataframes that contain only the corresponding genes
+subset_train_df = train_df.filter(gene_list, axis = 1)
+subset_test_df = test_df.filter(gene_list, axis = 1)
```
Comment (gwaybio) [inline 2.train-VAE/scripts/nbconverted/beta_vae_training.py:73]: This is ok for now to live here, but it really does belong in the `data_load()` functionality

File: 2.train-VAE/scripts/nbconverted/beta_vae_training.py [hunk 3]
```diff
@@ -29,56 +28,117 @@
...
 
+print(subset_test_df.shape)
+subset_test_df.head(3)
+
+
+# In[7]:
+
+
+# scale the data
+def absolute_maximum_scale(series):
```
Comment (gwaybio) [inline 2.train-VAE/scripts/nbconverted/beta_vae_training.py]: oh interesting, I don't remember talking about this. What is this normalization?

File: 2.train-VAE/scripts/nbconverted/beta_vae_training.py [hunk 4]
```diff
@@ -29,56 +28,117 @@
...
+
+
+# In[8]:
+
+
+encoder_architecture = []
+decoder_architecture = []
+
+
+# In[9]:
```
Comment (gwaybio) [inline 2.train-VAE/scripts/nbconverted/beta_vae_training.py]: Please add a note here saying where you discovered these optimal hyperparameters.

File: 2.train-VAE/scripts/nbconverted/beta_vae_training.py [hunk 5]
```diff
@@ -29,56 +28,117 @@
...
+
+
+cp_vae.train(x_train = subset_train_df, x_test = subset_test_df)
+
+
+# In[11]:
+
+
+# display training history
+history_df = pd.DataFrame(cp_vae.vae.history.history)
```
Comment (gwaybio) [inline 2.train-VAE/scripts/nbconverted/beta_vae_training.py:128]: You will want to output this file as well. Save it to a `results/` folder. Name it something like, "beta_vae_training_curve.csv"

File: 2.train-VAE/scripts/nbconverted/beta_vae_training.py [hunk 6]
```diff
@@ -29,56 +28,117 @@
...
+history_df
+
+
+# In[12]:
+
 
+# plot and save the figure
+save_path = pathlib.Path('../1.data-exploration/figures/training_curve.png')
```
Comment (gwaybio) [inline 2.train-VAE/scripts/nbconverted/beta_vae_training.py:139]: This is great - for when you make the final figure, outputting the training curve data in the cell above is optimal. This way you can modify the visualization without having to re-train each and every time :)

File: 2.train-VAE/scripts/nbconverted/beta_vae_training.py [hunk 7]
```diff
@@ -29,56 +28,117 @@
 # In[2]:
 
 
-# load the data 
-data_directory = "../0.data-download/data/"
-dfs = load_data(data_directory, adult_or_pediatric = "all")
-dependency_df = dfs[1]
-sample_df = dfs[0]
+# load the data
```
Comment (gwaybio) [inline 2.train-VAE/scripts/nbconverted/beta_vae_training.py:46]: In any notebooks or scripts that have stochastic elements, it is best to set a random seed at the top. https://www.geeksforgeeks.org/random-seed-in-python/

Please add a random seed to this notebook.

File: 2.train-VAE/scripts/nbconverted/fetch_hyper_params.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+data = [{'latent_dim': vis_data[idx]['hyperparameters']['values']['latent_dim'],
+         'learning_rate': vis_data[idx]['hyperparameters']['values']['learning_rate'], 
+         'beta': vis_data[idx]['hyperparameters']['values']['beta'], 
+         'encoder_batch_norm': vis_data[idx]['hyperparameters']['values']['encoder_batch_norm'], 
+         'batch_size': vis_data[idx]['hyperparameters']['values']['batch_size'],
+         'epochs': vis_data[idx]['hyperparameters']['values']['epochs'], 
+         'loss': vis_data[idx]['metrics']['metrics']['loss']['observations'][0]['value'],  
+         'val_loss': vis_data[idx]['metrics']['metrics']['val_loss']['observations'][0]['value'], } for idx in range(len(vis_datas[0]))]
+
+hip.Experiment.from_iterable(data).display()
```
Comment (gwaybio) [inline 2.train-VAE/scripts/nbconverted/fetch_hyper_params.py]: Please add the hiplot to this PR. Were you able to get the training_loss visualization working? This will actually be very important to figure out...

**Review by gwaybio — CHANGES_REQUESTED** [review]
Looks great @mjw-18 - I made a couple comments that you need to address prior to merging.

File: .gitignore (hunk 1 above)
Comment (mjw-18) [reply .gitignore]: this file was created as an outline for hyper parameterization but had not uses. i should have deleted it instead of using gitignore

File: 2.train-VAE/scripts/nbconverted/beta_vae_training.py (hunk 3 above)
Comment (mjw-18) [reply 2.train-VAE/scripts/nbconverted/beta_vae_training.py]: this is an absolute scaling method that sets all values to fall between 1 and -1

File: 2.train-VAE/scripts/nbconverted/beta_vae_training.py (hunk 4 above)
Comment (mjw-18) [reply 2.train-VAE/scripts/nbconverted/beta_vae_training.py]: Im going to add the note further up in the code so that it is next to the VAE parameter initiation

File: 2.train-VAE/scripts/nbconverted/beta_vae_training.py (hunk 3 above)
Comment (gwaybio) [reply 2.train-VAE/scripts/nbconverted/beta_vae_training.py]: gotcha - this normalization is not typical. Can you add a reference to the code to describe why you chose this method?

Comment (mjw-18) [reply 2.train-VAE/scripts/nbconverted/beta_vae_training.py]: this was a safety normalization to ensure that there were no outlying values in the data set. the actual changes to the data were minimal. Im beggining to think that this extra normalization code is a bit arbitrary

File: 2.train-VAE/scripts/nbconverted/fetch_hyper_params.py (hunk 8 above)
Comment (mjw-18) [reply 2.train-VAE/scripts/nbconverted/fetch_hyper_params.py]: https://github.com/facebookresearch/hiplot/issues/14#issuecomment-581047660

Comment (mjw-18) [reply 2.train-VAE/scripts/nbconverted/fetch_hyper_params.py]: Im still working on filtering through NaNs so that they will display

**Comment by mjw-18** [comment]
@gwaybio all updates to the codes are in!

File: 2.train-VAE/optimize.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        "--project_name",
+        architecture,
+        "--architecture",
+        architecture,
+    ]
+    all_commands.append(project_params)
+
+if __name__ == "__main__":
+    for cmd in all_commands:
+        subprocess.call(cmd)
```
Comment (gwaybio) [inline 2.train-VAE/optimize.py]:
```suggestion
        subprocess.call(cmd)

```

Adding a new line at the end of files to remove that red symbol

File: 2.train-VAE/optimize_hyperparameters.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+# create dataframe containing the 1000 genes with the largest variances and their corresponding gene label and extract the gene labels
+largest_var_df = gene_stats.nlargest(1000, "variance")
+gene_list = largest_var_df["gene_ID"].tolist()
+gene_list
+
+# create new training and testing dataframes that contain only the corresponding genes
+train_df = train_feat.filter(gene_list, axis = 1)
+test_df = test_feat.filter(gene_list, axis = 1)
+
+# scvale the data
```
Comment (gwaybio) [inline 2.train-VAE/optimize_hyperparameters.py]: I thought we talked about removing this

File: 2.train-VAE/optimize_utils.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+        default="onelayer", 
+        help="VAE architecture")
+    
+    parser.add_argument(
+        "--dataset", 
+        default="cell-painting", 
+        help="cell-painting or L1000 dataset")
+    
+    args = parser.parse_args()
+    return args
```
Comment (gwaybio) [inline 2.train-VAE/optimize_utils.py]:
```suggestion
    return args

```

New line

File: 2.train-VAE/scripts/nbconverted/beta_vae_training.py [hunk 12]
```diff
@@ -8,77 +8,259 @@
...
+
+
+# plot and save the figure
+save_path = pathlib.Path('../1.data-exploration/figures/training_curve.png')
+
+plt.figure(figsize=(6, 5), dpi = 500)
+plt.plot(history_df["loss"], label="Training data")
+plt.plot(history_df["val_loss"], label="Validation data")
+plt.ylabel("MSE + KL Divergence")
+plt.xlabel("No. Epoch")
```
Comment (gwaybio) [inline 2.train-VAE/scripts/nbconverted/beta_vae_training.py]:
```suggestion
plt.xlabel("Epoch")
```

The `No.` is implied by the axis tick labels.

File: 2.train-VAE/scripts/nbconverted/beta_vae_training.py [hunk 13]
```diff
@@ -8,77 +8,259 @@
...
+# In[18]:
+
+
+# create a data frame of both test and train gene dependency data sorted by top 1000 highest gene variances
+concat_frames = [train_init, test_init]
+train_and_test = pd.concat(concat_frames).reset_index(drop=True)
+train_and_test[["age_category", "sex"]] = train_and_test.age_and_sex.str.split(pat='_',expand=True)
+train_and_test_subbed = train_and_test.filter(gene_list, axis = 1)
+asdf = []
+asdf = pd.DataFrame(asdf)
```
Comment (gwaybio) [inline 2.train-VAE/scripts/nbconverted/beta_vae_training.py]: can you name the variable something more descriptive?

File: 2.train-VAE/scripts/nbconverted/beta_vae_training.py [hunk 14]
```diff
@@ -8,77 +8,259 @@
...
+data_dir = "../0.data-download/data/"
+datafs = load_data(data_dir, adult_or_pediatric = "all")
+dependency_df = datafs[1]
+sample_df = datafs[0]
+
+
+# In[17]:
+
+
+train_init['train_or_test'] = train_init.apply(lambda _: 'train', axis=1)
```
Comment (gwaybio) [inline 2.train-VAE/scripts/nbconverted/beta_vae_training.py:177]: what is this doing? I cannot tell

File: 2.train-VAE/scripts/nbconverted/beta_vae_training.py [hunk 15]
```diff
@@ -8,77 +8,259 @@
...
+    train_or_test = train_and_test.train_or_test.astype(str)
+    )
+metadata
+
+
+# In[19]:
+
+
+latent_complete = np.array(encoder.predict(train_and_test_subbed)[2])
+latent_df = pd.DataFrame(latent_complete)
```
Comment (gwaybio) [inline 2.train-VAE/scripts/nbconverted/beta_vae_training.py:207]: Please output `latent_df` to a file in the results folder as well.

File: 2.train-VAE/scripts/nbconverted/beta_vae_training.py [hunk 16]
```diff
@@ -8,77 +8,259 @@
...
+    zz.append(z)
+# add the third legend
+legend5 = plt.legend(zz, train_test.unique(), loc="center", title='train or test', ncol=2, bbox_to_anchor=(1.3, .45), bbox_transform=gcf().transFigure)
+plt.gca().add_artist(legend4)
 
 
-cp_vae.train(x_train=dependency_df, x_test=dependency_df)
+# save the figure
+heat_save_path = pathlib.Path('../1.data-exploration/figures/heatmap.png')
+g.figure.savefig(heat_save_path, dpi=600)
```
Comment (gwaybio) [inline 2.train-VAE/scripts/nbconverted/beta_vae_training.py]: I don't see legends in this figure that you're adding to the PR?
[
1.data-exploration/figures/heatmap.png](https://github.com/WayScience/gene_dependency_representations/blob/f7f5585a426fc12bc5d346f4fceee72d448a1bb8/1.data-exploration/figures/heatmap.png)

Is this an old file?

File: 2.train-VAE/utils.py [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+        features = [
+        x
+        for x in population_df.columns.tolist()
+        if (
+            not x.isnumeric()
+        )
+    ]
+    
+    return features
+
```
Comment (gwaybio) [inline 2.train-VAE/utils.py]:
```suggestion

```

File: 2.train-VAE/vae_utils.py [hunk 18]
```diff
@@ -114,3 +114,4 @@ def connect_decoder(input_dim, latent_dim, architecture=[]):
     nodes["outputs"] = Dense(input_dim, activation='linear')(nodes[idx])
 
     nodes["decoder"] = Model(nodes["inputs"], nodes["outputs"])
+    return nodes
```
Comment (gwaybio) [inline 2.train-VAE/vae_utils.py]:
```suggestion
    return nodes

```

**Review by gwaybio — CHANGES_REQUESTED** [review]
A couple more minor comments. Please address them, and then I will review and merge

File: 2.train-VAE/optimize_hyperparameters.py (hunk 10 above)
Comment (mjw-18) [reply 2.train-VAE/optimize_hyperparameters.py]: This is an important line. It filters out the 1000 genes with the highest variances that are then used for hyperparameterization

Comment (mjw-18) [reply 2.train-VAE/optimize_hyperparameters.py]: ohhh haha you meant the line under it. I removed that normalization function just now. I didn't realize I added it to two places

File: 2.train-VAE/scripts/nbconverted/beta_vae_training.py (hunk 14 above)
Comment (mjw-18) [reply 2.train-VAE/scripts/nbconverted/beta_vae_training.py:177]: These functions add a new column to a data frame and then add the designated string to every cell in the newly created column. The intent of this new column is to add labels to ID which were training and testing samples so they could be labeled in the heat map.

File: 2.train-VAE/utils.py (hunk 17 above)
Comment (mjw-18) [reply 2.train-VAE/utils.py]: why should this .py file not end with a new line?

File: 2.train-VAE/scripts/nbconverted/beta_vae_training.py (hunk 16 above)
Comment (mjw-18) [reply 2.train-VAE/scripts/nbconverted/beta_vae_training.py]: it was an issue with the save command. its been resolved and wil push it with the other updates

File: 2.train-VAE/utils.py (hunk 17 above)
Comment (gwaybio) [reply 2.train-VAE/utils.py]: it should. This suggestion is to delete 4 extra spaces

Comment (mjw-18) [reply 2.train-VAE/utils.py]: oh okay

**Comment by mjw-18** [comment]
@gwaybio sweet! all the updates should be in!

**Review by gwaybio — APPROVED** [review]
Looks great! Please merge when you're ready

(also, make sure that after you merge you follow [step 7](https://github.com/PennGCB/sandbox#step-7---reconfigure-local-machine-and-fork-important))
